File: frontend/src/engine/capabilities/insertTextBox.ts
/**
 * insertTextBox – Insert a text box into a worksheet.
 *
 * Office.js notes:
 * - worksheet.shapes.addTextBox(text) creates a text box shape.
 * - Position and size via shape.left/top/width/height in pixels.
 * - Font properties via shape.textFrame.textRange.font.
 * - Fill via shape.fill.
 */

import { CapabilityMeta, InsertTextBoxParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";

const meta: CapabilityMeta = {
  action: "insertTextBox",
  description: "Insert a text box with custom text and formatting",
  mutates: true,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.9",
};

async function handler(
  context: Excel.RequestContext,
  params: InsertTextBoxParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { sheetName, text, left, top, width, height, fontSize, fontFamily, fontColor, fillColor, horizontalAlignment } = params;

  if (options.dryRun) {
    const preview = text.length > 30 ? text.substring(0, 30) + "..." : text;
    return {
      stepId: "",
      status: "success",
      message: `Would insert text box "${preview}" at (${left}, ${top})`,
    };
  }

  options.onProgress?.("Inserting text box...");

  // Resolve target sheet
  let sheet: Excel.Worksheet;
  if (sheetName) {
    const ws = context.workbook.worksheets.getItemOrNullObject(sheetName);
    ws.load("isNullObject");
    await context.sync();
    if (ws.isNullObject) {
      return {
        stepId: "",
        status: "error",
        message: `Sheet "${sheetName}" not found. Please check the sheet name.`,
      };
    }
    sheet = ws;
  } else {
    sheet = context.workbook.worksheets.getActiveWorksheet();
  }

  const shape = sheet.shapes.addTextBox(text);

  shape.left = left;
  shape.top = top;
  shape.width = width;
  shape.height = height;

  // Font properties
  const font = shape.textFrame.textRange.font;
  if (fontSize !== undefined) font.size = fontSize;
  if (fontFamily !== undefined) font.name = fontFamily;
  if (fontColor !== undefined) font.color = fontColor;

  // Fill
  if (fillColor) {
    shape.fill.setSolidColor(fillColor);
  }

  // Horizontal alignment
  if (horizontalAlignment) {
    const alignMap: Record<string, Excel.ShapeTextHorizontalAlignment> = {
      left: Excel.ShapeTextHorizontalAlignment.left,
      center: Excel.ShapeTextHorizontalAlignment.center,
      right: Excel.ShapeTextHorizontalAlignment.right,
    };
    const align = alignMap[horizontalAlignment];
    if (align !== undefined) {
      shape.textFrame.horizontalAlignment = align;
    }
  }

  await context.sync();

  const preview = text.length > 30 ? text.substring(0, 30) + "..." : text;
  return {
    stepId: "",
    status: "success",
    message: `Inserted text box "${preview}" at (${left}, ${top})${sheetName ? ` on "${sheetName}"` : ""}`,
  };
}

// ── Legacy-Excel fallback (ExcelApi < 1.9) ────────────────────────────────────
// TextBox shapes require 1.9. We substitute a merged cell block with centered
// wrapped text, optional fill color, and a light border. Fidelity cost: no
// free positioning (snaps to the grid), no auto-size-to-content, no overlay
// over other cells. Most annotation/label use cases survive this intact.
async function fallback(
  context: Excel.RequestContext,
  params: InsertTextBoxParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const {
    sheetName, text, left, top, width, height,
    fontSize, fontFamily, fontColor, fillColor, horizontalAlignment,
  } = params;

  if (options.dryRun) {
    const preview = text.length > 30 ? text.substring(0, 30) + "..." : text;
    return {
      stepId: "",
      status: "success",
      message: `Would emit merged-cell textbox "${preview}" (legacy fallback).`,
    };
  }

  options.onProgress?.("Legacy-Excel mode: faking textbox via merged cells...");

  const sheet = sheetName
    ? context.workbook.worksheets.getItem(sheetName)
    : context.workbook.worksheets.getActiveWorksheet();

  // Approximate column/row size (Excel defaults).
  const COL_PT = 64;
  const ROW_PT = 15;
  const startCol = Math.max(0, Math.round(left / COL_PT));
  const startRow = Math.max(0, Math.round(top / ROW_PT));
  const colSpan = Math.max(1, Math.round(width / COL_PT));
  const rowSpan = Math.max(1, Math.round(height / ROW_PT));

  const block = sheet.getRangeByIndexes(startRow, startCol, rowSpan, colSpan);
  if (rowSpan > 1 || colSpan > 1) {
    try { block.merge(true); } catch { /* may already be merged */ }
  }

  // Write the text into the (now merged) top-left cell. values on a merged
  // range is a single-cell grid, but Office.js expects a rowSpan×colSpan 2D
  // for range.values. After merge the range reports as the merged block; feed
  // a 2D array of matching shape where only [0][0] has the text.
  const grid: (string | null)[][] = [];
  for (let r = 0; r < rowSpan; r++) {
    const row: (string | null)[] = [];
    for (let c = 0; c < colSpan; c++) row.push(r === 0 && c === 0 ? text : null);
    grid.push(row);
  }
  try {
    block.values = grid as unknown as (string | number | boolean)[][];
  } catch {
    // After merge Office.js may prefer a 1×1 write to the merged cell.
    sheet.getRangeByIndexes(startRow, startCol, 1, 1).values = [[text]];
  }

  // Font
  if (fontSize !== undefined) block.format.font.size = fontSize;
  if (fontFamily !== undefined) block.format.font.name = fontFamily;
  if (fontColor !== undefined) block.format.font.color = fontColor;

  // Fill
  if (fillColor) block.format.fill.color = fillColor;

  // Alignment — default to center so the merged block looks like a textbox.
  const alignMap: Record<string, Excel.HorizontalAlignment> = {
    left:   Excel.HorizontalAlignment.left,
    center: Excel.HorizontalAlignment.center,
    right:  Excel.HorizontalAlignment.right,
  };
  block.format.horizontalAlignment = alignMap[horizontalAlignment ?? "center"] ?? Excel.HorizontalAlignment.center;
  block.format.verticalAlignment = Excel.VerticalAlignment.center;
  block.format.wrapText = true;

  // Border to visually separate the "textbox" from surrounding cells.
  const sides: Excel.BorderIndex[] = [
    Excel.BorderIndex.edgeTop,
    Excel.BorderIndex.edgeBottom,
    Excel.BorderIndex.edgeLeft,
    Excel.BorderIndex.edgeRight,
  ];
  for (const side of sides) {
    const b = block.format.borders.getItem(side);
    b.style = Excel.BorderLineStyle.continuous;
    b.weight = Excel.BorderWeight.thin;
    b.color = "#808080";
  }

  await context.sync();

  const preview = text.length > 30 ? text.substring(0, 30) + "..." : text;
  return {
    stepId: "",
    status: "success",
    message:
      `Inserted textbox approximation "${preview}" at row ${startRow + 1}, col ${startCol + 1} ` +
      `(${rowSpan}×${colSpan} merged block) — textbox shapes require ExcelApi 1.9+ ` +
      `(legacy-Excel fallback; snaps to cell grid, no free positioning).`,
  };
}

registry.register(meta, handler as any, fallback as any);
export { meta };

File: backend/app/execution/capabilities/insert_text_box.py
"""
insertTextBox — insert a text box via the COM Shapes.AddTextbox API.

Styling mirrors the TS handler: font (size/family/color), fill color, and
horizontal alignment. Orientation=1 = horizontal text flow.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry


# msoTextOrientationHorizontal = 1
_ORIENT_HORIZONTAL = 1

# XlHAlign-style alignment values for TextFrame2.TextRange.ParagraphFormat.Alignment.
# msoAlignLeft=1, msoAlignCenter=2, msoAlignRight=3
_ALIGN_MAP = {
    "left": 1,
    "center": 2,
    "centre": 2,
    "right": 3,
}


def _hex_to_rgb_int(color: str) -> int:
    h = color.lstrip("#")
    if len(h) != 6:
        raise ValueError(f"Invalid color {color!r}; expected #RRGGBB.")
    return int(h[0:2], 16) + int(h[2:4], 16) * 256 + int(h[4:6], 16) * 65536


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    text = params.get("text")
    left = params.get("left")
    top = params.get("top")
    width = params.get("width")
    height = params.get("height")
    sheet_name = params.get("sheetName")
    font_size = params.get("fontSize")
    font_family = params.get("fontFamily")
    font_color = params.get("fontColor")
    fill_color = params.get("fillColor")
    h_align = params.get("horizontalAlignment")

    if text is None or left is None or top is None or width is None or height is None:
        return {
            "status": "error",
            "message": "insertTextBox requires 'text', 'left', 'top', 'width', 'height'.",
        }

    preview = text if len(text) <= 30 else text[:30] + "..."

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would insert text box {preview!r} at ({left}, {top}).",
        }

    book = ctx.workbook_handle
    try:
        if sheet_name:
            try:
                sheet = book.sheets[sheet_name]
            except Exception:
                return {
                    "status": "error",
                    "message": f"Sheet {sheet_name!r} not found. Please check the sheet name.",
                }
        else:
            sheet = book.sheets.active

        shape = sheet.api.Shapes.AddTextbox(
            _ORIENT_HORIZONTAL,
            float(left),
            float(top),
            float(width),
            float(height),
        )

        # Set text first, then styling.
        try:
            shape.TextFrame2.TextRange.Text = str(text)
        except Exception:
            shape.TextFrame.Characters.Text = str(text)

        # Font styling — prefer TextFrame2 (modern), fall back to Characters.Font.
        try:
            font = shape.TextFrame2.TextRange.Font
            if font_size is not None:
                font.Size = float(font_size)
            if font_family is not None:
                font.Name = str(font_family)
            if font_color is not None:
                font.Fill.ForeColor.RGB = _hex_to_rgb_int(font_color)
        except Exception:  # noqa: BLE001 — fall back for older Excel
            try:
                font = shape.TextFrame.Characters.Font
                if font_size is not None:
                    font.Size = float(font_size)
                if font_family is not None:
                    font.Name = str(font_family)
                if font_color is not None:
                    font.Color = _hex_to_rgb_int(font_color)
            except Exception:  # noqa: BLE001
                pass

        if fill_color:
            try:
                shape.Fill.ForeColor.RGB = _hex_to_rgb_int(fill_color)
                shape.Fill.Solid()
            except Exception:  # noqa: BLE001
                pass

        if h_align:
            align_val = _ALIGN_MAP.get(h_align.lower() if isinstance(h_align, str) else "")
            if align_val is not None:
                try:
                    shape.TextFrame2.TextRange.ParagraphFormat.Alignment = align_val
                except Exception:
                    # xlHAlign on TextFrame: xlHAlignLeft=-4131, center=-4108, right=-4152
                    legacy = {1: -4131, 2: -4108, 3: -4152}.get(align_val)
                    if legacy is not None:
                        try:
                            shape.TextFrame.HorizontalAlignment = legacy
                        except Exception:  # noqa: BLE001
                            pass
    except Exception as exc:  # noqa: BLE001
        return {
            "status": "error",
            "message": f"Insert text box failed: {exc}",
            "error": str(exc),
        }

    on_sheet = f" on {sheet_name!r}" if sheet_name else ""
    return {
        "status": "success",
        "message": f"Inserted text box {preview!r} at ({left}, {top}){on_sheet}.",
        "outputs": {"sheet": sheet.name},
    }


registry.register("insertTextBox", handler, mutates=True)
